File: craft/commands/ssh_keys.py
import click

from craft.client import delete, get, post
from craft.output import print_item, print_json, print_success, print_table


@click.command("list")
def sshkey_list():
    """List saved SSH keys."""
    data = get("/ssh-keys")
    items = data.get("data", data)
    if isinstance(items, dict):
        items = items.get("items", items.get("keys", []))
    if isinstance(items, list) and items:
        rows = []
        for k in items:
            rows.append([
                k.get("id", ""),
                k.get("name", ""),
                k.get("fingerprint", k.get("publicKey", "")[:40]),
            ])
        print_table(rows, ["ID", "Name", "Fingerprint"])
    else:
        print_json(data)


@click.command("add")
@click.option("--name", required=True, help="Key name (1-100 chars)")
@click.option("--public-key", required=True, help="OpenSSH public key")
def sshkey_add(name, public_key):
    """Add an SSH public key."""
    data = post("/ssh-keys", {"name": name, "publicKey": public_key})
    print_success("SSH key added.")
    print_item(data)


@click.command("add-file")
@click.option("--name", required=True, help="Key name (1-100 chars)")
@click.option("--file", "key_file", required=True, type=click.Path(exists=True), help="Path to public key file")
def sshkey_add_file(name, key_file):
    """Add an SSH key from a file."""
    with open(key_file) as f:
        public_key = f.read().strip()
    data = post("/ssh-keys", {"name": name, "publicKey": public_key})
    print_success("SSH key added.")
    print_item(data)


@click.command("delete")
@click.argument("key_id")
def sshkey_delete(key_id):
    """Delete an SSH key."""
    delete(f"/ssh-keys/{key_id}")
    print_success(f"SSH key {key_id} deleted.")

File: cmd/ssh_keys.go
package cmd

import (
	"fmt"
	"os"

	"github.com/CraftIntertech/craft-cli/internal/client"
	"github.com/CraftIntertech/craft-cli/internal/output"
	"github.com/spf13/cobra"
)

var sshKeyListCmd = &cobra.Command{
	Use:   "list",
	Short: "List all SSH keys",
	Run: func(cmd *cobra.Command, args []string) {
		result := client.Get("/ssh-keys", nil, true)
		keys := extractList(result, "ssh_keys", "sshKeys")

		var rows [][]string
		for _, k := range keys {
			rows = append(rows, []string{
				getString(k, "id"),
				getString(k, "name"),
				getString(k, "fingerprint"),
			})
		}

		output.PrintTable(rows, []string{"ID", "Name", "Fingerprint"})
	},
}

var sshKeyAddCmd = &cobra.Command{
	Use:   "add",
	Short: "Add a new SSH key",
	Run: func(cmd *cobra.Command, args []string) {
		name, _ := cmd.Flags().GetString("name")
		publicKey, _ := cmd.Flags().GetString("public-key")

		if name == "" {
			name = PromptText("SSH key name:")
		}
		if publicKey == "" {
			publicKey = PromptText("Public key:")
		}

		body := map[string]interface{}{
			"name":      name,
			"publicKey": publicKey,
		}

		result := client.Post("/ssh-keys", body, true, 0)
		output.PrintSuccess("SSH key added.")
		output.PrintItem(extractObj(result))
	},
}

var sshKeyAddFileCmd = &cobra.Command{
	Use:   "add-file",
	Short: "Add a new SSH key from a file",
	Run: func(cmd *cobra.Command, args []string) {
		name, _ := cmd.Flags().GetString("name")
		filePath, _ := cmd.Flags().GetString("file")

		if name == "" {
			name = PromptText("SSH key name:")
		}
		if filePath == "" {
			filePath = PromptText("Path to public key file:")
		}

		data, err := os.ReadFile(filePath)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error reading file: %v\n", err)
			os.Exit(1)
		}

		body := map[string]interface{}{
			"name":      name,
			"publicKey": string(data),
		}

		result := client.Post("/ssh-keys", body, true, 0)
		output.PrintSuccess("SSH key added from file.")
		output.PrintItem(extractObj(result))
	},
}

var sshKeyDeleteCmd = &cobra.Command{
	Use:   "delete [KEY_ID]",
	Short: "Delete an SSH key",
	Args:  cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		keyID := args[0]

		if !Confirm(fmt.Sprintf("Are you sure you want to delete SSH key %s?", keyID)) {
			output.PrintInfo("Cancelled.")
			return
		}

		client.Delete(fmt.Sprintf("/ssh-keys/%s", keyID), true)
		output.PrintSuccess("SSH key deleted.")
	},
}

func init() {
	sshKeyAddCmd.Flags().String("name", "", "Name for the SSH key")
	sshKeyAddCmd.Flags().String("public-key", "", "Public key content")

	sshKeyAddFileCmd.Flags().String("name", "", "Name for the SSH key")
	sshKeyAddFileCmd.Flags().String("file", "", "Path to public key file")

	SSHKeyCmd.AddCommand(sshKeyListCmd)
	SSHKeyCmd.AddCommand(sshKeyAddCmd)
	SSHKeyCmd.AddCommand(sshKeyAddFileCmd)
	SSHKeyCmd.AddCommand(sshKeyDeleteCmd)
}
